app/Http/Controllers/Auth/OAuthController.php

```diff
@@ -44,47 +44,97 @@ public function redirect($provider)
-        $user = $this->findOrCreateUser($provider, $user);
-
-        $this->guard()->setToken(
-            $token = $this->guard()->login($user)
-        );
-
-        return view('oauth/callback', [
-            'token' => $token,
-            'token_type' => 'bearer',
-            'expires_in' => $this->guard()->getPayload()->get('exp') - time(),
-        ]);
+        try {
+            $driverUser = $provider->getDriver()->setRedirectUrl(config('services.google.auth_redirect'))->getUser();
+            $user = $this->findOrCreateUser($provider, $driverUser);
+
+            if (!$user) {
+                return $this->error([
+                    "message" => "User not found."
+                ]);
+            }
+
+            if ($user->has_registered) {
+                return $this->error([
+                    "message" => "This email is already registered. Please sign in with your password."
+                ]);
+            }
+
+            $this->guard()->setToken(
+                $token = $this->guard()->login($user)
+            );
+
+            return response()->json([
+                'token' => $token,
+                'token_type' => 'bearer',
+                'expires_in' => $this->guard()->getPayload()->get('exp') - time(),
+                'new_user' => $user->new_user
+            ]);
+        } catch (\Exception $e) {
+            return $this->error([
+                "message" => "OAuth service failed to authenticate: " . $e->getMessage()
+            ]);
+        }
     }
 
     /**
-     * @param  string  $provider
-     * @param  \Laravel\Socialite\Contracts\User  $sUser
-     * @return \App\Models\User
+     * @p   aram  \Laravel\Socialite\Contracts\User  $socialiteUser
+     * @return \App\Models\User | null
      */
-    protected function findOrCreateUser($provider, $user)
+    protected function findOrCreateUser($provider, $socialiteUser)
     {
         $oauthProvider = OAuthProvider::where('provider', $provider)
-            ->where('provider_user_id', $user->getId())
+            ->where('provider_user_id', $socialiteUser->getId())
             ->first();
 
         if ($oauthProvider) {
             $oauthProvider->update([
-                'access_token' => $user->token,
-                'refresh_token' => $user->refreshToken,
+                'access_token' => $socialiteUser->token,
+                'refresh_token' => $socialiteUser->refreshToken,
             ]);
 
             return $oauthProvider->user;
         }
 
-        if (User::where('email', $user->getEmail())->exists()) {
-            throw new EmailTakenException();
+
+        if (!$provider->getDriver()->canCreateUser()) {
+            return null;
         }
 
-        return $this->createUser($provider, $user);
+        $email = strtolower($socialiteUser->getEmail());
+        $user = User::whereEmail($email)->first();
+        if ($user) {
+            $user->has_registered = true;
+            return $user;
+        } else {
```

Comment:
No need for the `else` statement, you can just put the code inline since there's a return statement